### DISCUSSION TITLE
Security issues in BLS Signatures

### DISCUSSION THREAD

**Discussion by gimer** [cartesi/cips#27] [General]
Security issues in BLS Signatures

This topic intends to keep track of vulnerabilities and weaknesses of BLS signatures. The BLS digital signature scheme allows to aggregate a set of signatures to save storage space, bandwidth or processing time aiming to reduce gas fees. There are several implementations based on the draft [1], however there might be vulnerabilities to take into account so that we can make informed decisions. The final goal is of the following attacks is to compromise the non-repudiation property of this signatures scheme: 

1. Zero bugs and Splitting zero attacks [2]: 

+ Zero bugs attacks referee to a particular property of BLS signatures, if the public key is equal to zero the private and the signatures are equal to zero. From the verifiers perspective the signature is invalid because it is not possible to know what message was originally signed. The BLS RFC draft v4 warns about checking zero public keys but this does not prevent from an inaccurate implementation. For instance, the implementation: hermit/bls [4], did not implement zero public keys validations in its original design. According of the author, Ethereum’s py_ecc [3] checks for zero but the verification is not accurate. 

+ Splitting Zero attack refers to an attack when a pair of public keys ( X and Y) both different from zero and their related digital signatures (d1, d2) also both different from zero such that: X + Y = 0 and d1 + d2 = 0.  In plain words, if the attacker manage to craft a pair of public keys different from zero such that the combination is equal to zero and their related digital signatures also different from zero but the combination is equal to zero, then the result of the functions: Verify and AggregateVerify can be compromised. These functions are implemented according to the BLS RFC draft v4 [1]. According of the author, the implementations: py_cc [3] , blst [5] and milagro_bls [6], all have the same bug. 

2. Consensus attacks: the attacker’s goal is disrupt the global view of a set of signatures by creating a set of invalid individual signatures such that their aggregate signature is valid. Some users will see valid signatures while others  will see invalid signatures. As a result, two equivalent interfaces might return conflicting results: FastAggregateVerify and AggregateVerify. The attack described in [7] has been fully documented by Jan Camenish et. al. in the paper: Batch verification of short signatures. This paper was published 14 years ago, however the possible impact in blockchain technology is still work in progress.

[1] BLS Signatures Draft v4 https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04

[2] https://eprint.iacr.org/2021/323.pdf

[3] https://github.com/ethereum/py_ecc

[4] https://github.com/herumi/bls

[5] https://github.com/supranational/blst

[6] https://github.com/sigp/milagro_bls

[7] https://eprint.iacr.org/2021/377.pdf

**Comment by augustoteixeira** [comment]
This is a great research, Gimer!

One thing that got me really worried about was the "Splitting Zeros" attack. It game me two impressions:
- perhaps it is possible to make similar attacks (like X + Y + Z = 0) and it could be quite difficult to identify if there is a set of keys inside the block that sum up to zero.
- perhaps the whole BLS is still at its infancy and we should wait more to see it mature?

I really don't know. Let us keep investigating and also give ZK alternatives a look. What do you think?

**Reply by gimer** [reply]
Thank you @augustoteixeira, 

As you mentioned, I also have the impression that there is a possibility of similar attacks. I think BLS signatures are secure, however the implementations are not accurate. So, we need to be careful if we choose one of them. I also think that the implications of the attacks in the Blockchain space are unknown at this moment.

I'll keep looking and also exploring ZK approaches as well ( e.g., cairo).
